doc/features/open-telemetry/tracing-rfc.md

```diff
@@ -0,0 +1,413 @@
+they should only be used if the library being instrumented already provides them. When it's otherwise impossible to get any meaningful span name, `db.name` or the tech-specific database name MAY be used."
+
+To avoid parsing the statement, the **span name** in this implementation has the **keyspace name** as its value.
+
+### Span attributes
+
+This implementation will include, by default, the **required** attributes for Database, and Cassandra spans.\
+`server.address` and `server.port`, despite only **recommended**, are included to give information regarding the client connection.\
+`db.statement` is optional given that this attribute may contain sensitive information:
+
+| Attribute  | Description  | Type | Required | Supported Values|
+|---|---|---|---|---|
+| span.kind | Describes the relationship between the Span, its parents, and its children in a Trace. | string | true | client |
+| db.system | An identifier for the database management system (DBMS) product being used. | string | true | cassandra |
+| db.name | The keyspace name in Cassandra. | string | true | *keyspace in use* |
+| db.statement | The database statement being executed. | string | false | *database statement in use* |
+| server.address | Name of the database host. | string | true | e.g.: example.com; 10.1.2.80; /tmp/my.sock |
+| server.port | Server port number. Used in case the port being used is not the default. | int | false | e.g.: 9445 |
+
+## Usage
+
+### Package installation
+
+The OpenTelemetry implementation will be included in the package `CassandraCSharpDriver.OpenTelemetry`.
+
+### Exporting Cassandra activity
+
+The extension method `AddOpenTelemetryTraceInstrumentation()` will be available in the cluster builder, so the activity can be exported for database operations:
+
+```csharp
+var cluster = Cluster.Builder()
+    .AddContactPoint("127.0.0.1")
+    .WithSessionName("session-name")
+    .AddOpenTelemetryTraceInstrumentation()
+    .Build();
+```
+
+The extension method also includes the option to enable the database statement that is disabled by default:
+
+```csharp
+var cluster = Cluster.Builder()
+    .AddContactPoint("127.0.0.1")
+    .WithSessionName("session-name")
+    .AddOpenTelemetryTraceInstrumentation(options => options.SetDbStatement = true)
+    .Build();
+```
+
+### Capturing Cassandra activity
+
+When setting up the tracer provider, it is necessary to include the Cassandra source for the activity to be captured. The activity source name will be available in the property `CassandraInstrumentation.ActivitySourceName` that will be included in the `CassandraCSharpDriver.OpenTelemetry` package.
+
+Example:
+
+```csharp
+using var tracerProvider = Sdk.CreateTracerProviderBuilder()
+ .AddSource(CassandraInstrumentation.ActivitySourceName)
+ .AddConsoleExporter()
+ .Build();
+```
```

Comment:
I'm a bit concerned about the correlation (which IIUC relies on `Activity.Current`). I think that no extra work is needed for "session" level tracing of requests which is effectively what this document proposes but we still need to verify this with testing.

For node level tracing (in the future) we might need extra work because I'm fairly certain the `OperationObserver` methods are being called with a different async/task context. It will also need additional work because I don't think we have access to the `IStatement` objects at that level.

The work involved in getting access to `IStatement` objects at the connection level is something that does not require any breaking changes to the API (so we don't have to worry about now) but getting access to the original async/task context (i.e. `Activity`) might need some changes to the API... Maybe adding a new method `object BuildContext` to the `IRequestTracker`/`IDriverTrace` interface and the driver would obtain a context `object` from this method call and provide that object when calling the other request tracker methods?

Reply:
Another possibility would be to have the driver pass the `Activity` to the request tracker but this would require the core package to have an additional dependency to be able to get the current `Activity` right?

Reply:
Hmm maybe I'm looking at this the wrong way. We don't actually need these methods to be called at the connection level, they can be called at the `RequestHandler`/`RequestExecution` level but some changes have to be made internally because there is no `OnNodeSuccess` and `OnNodeError` methods on the `IRequestObserver` interface.

Currently the `RequestHandler` and `RequestExecution` classes aren't calling any "observation" method whenever a response is actually received by a single node except for when an error is received.

Reply:
Added a revised proposal [here](https://github.com/datastax/csharp-driver/pull/603#discussion_r1478741930).

Reply:
I'm gonna close this discussion so we can focus on the `IRequestTracker` [discussion](https://github.com/datastax/csharp-driver/pull/603#discussion_r1478741930) to avoid going back and forth :)